core/src/components/button/button.ionic.scss

```diff
@@ -264,8 +272,47 @@
   @include globals.margin-horizontal(globals.$ion-space-300, null);
 }
 
-// Button Badge
+// Button with Badge
 // --------------------------------------------------
-:host ::slotted(ion-badge[vertical]:not(:empty)) {
-  @include globals.padding(globals.$ion-space-050);
+
+:host(.button-has-badge) {
+  --padding-top: #{globals.$ion-space-0};
+  --padding-bottom: #{globals.$ion-space-0};
+}
+
+:host(.button-small) ::slotted(ion-badge) {
+  @include globals.position(null, calc(-1 * globals.$ion-space-050), null, null);
+}
+
+:host(.button-medium),
+:host:not(.button-small):not(.button-large) {
```

Comment:
Why do we need this check? We are already targeting `medium`? Plus we try to minimize the use of `:not` when possible.